src/FastMathFunctions.cpp

```diff
@@ -0,0 +1,872 @@
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    return true;
+}
+
+bool native_func_satisfies_precision(const IntrinsicsInfo &ii, const ApproximationPrecision &prec) {
+    if (prec.force_halide_polynomial) {
+        return false;  // Don't use native functions if the user really wants a polynomial.
+    }
+    if (!ii.native_func.defined()) {
+        return true;  // Unspecified means it's exact.
+    }
+    if (prec.optimized_for != ii.native_func.behavior) {
+        return false;
+    }
+    if (prec.constraint_max_ulp_error != 0) {
+        if (ii.native_func.max_ulp_error != 0) {
+            if (ii.native_func.max_ulp_error > prec.constraint_max_ulp_error) {
+                return false;
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    if (prec.constraint_max_absolute_error != 0) {
+        if (ii.native_func.max_abs_error != 0) {
+            if (ii.native_func.max_abs_error > prec.constraint_max_absolute_error) {
+                return false;
+            }
+        } else {
+            // We don't know?
+        }
+    }
+    return true;
+}
+
+class LowerFastMathFunctions : public IRMutator {
+    using IRMutator::visit;
+
+    const Target &target;
+    DeviceAPI for_device_api = DeviceAPI::None;
+
+    bool is_cuda_cc20() {
+        return for_device_api == DeviceAPI::CUDA && target.get_cuda_capability_lower_bound() >= 20;
+    }
+    bool is_cuda_cc75() {
+        return for_device_api == DeviceAPI::CUDA && target.get_cuda_capability_lower_bound() >= 75;
+    }
+
+    void adjust_precision_for_target(ApproximationPrecision &prec) {
+        if (for_device_api == DeviceAPI::None) {
+            if (target.arch == Target::Arch::X86) {
+                // If we do not have fused-multiply-add, we lose some precision.
+                if (target.bits == 32 || !target.has_feature(Target::Feature::FMA)) {
+                    prec.constraint_max_absolute_error *= 0.5f;
+                    prec.constraint_max_ulp_error /= 2;
+                }
+            }
+        }
+    }
+
+    /** Strips the fast_ prefix, appends the type suffix, and
+     * drops the precision argument from the end. */
+    Expr to_native_func(const Call *op) {
+        internal_assert(op->name.size() > 5);
+        internal_assert(op->name.substr(0, 5) == "fast_");
+        internal_assert(op->args.size() >= 2);  // At least one arg, and a precision
+        std::string new_name = op->name.substr(5);
+        if (op->type == Float(16)) {
+            new_name += "_f16";
+        } else if (op->type == Float(32)) {
+            new_name += "_f32";
+        } else if (op->type == Float(64)) {
+            new_name += "_f64";
+        }
+        // Mutate args, and drop precision parameter.
+        std::vector<Expr> args;
+        for (size_t i = 0; i < op->args.size() - 1; ++i) {
+            const Expr &arg = op->args[i];
+            args.push_back(IRMutator::mutate(arg));
```

Comment:
Why is this IRMutator::mutate as opposed to just mutate?

Reply:
Fixed, thanks! Well spotted!